### ISSUE TITLE
Lifecycle hooks don't support methods

### ISSUE THREAD

**Issue by peterschutt** [starlite-api/starlite#284]
The docs do use the word `function` explicitly for `before_request` and `after_request` so based on that this probably isn't a bug, but bringing it up for sake of discussion.

```python
import time
from typing import Any

from starlite import Starlite, get


class Timer:
    def __init__(self) -> None:
        self._start = 0.0

    def start(self, _: Any) -> None:
        self._start = time.time()

    def end(self, _: Any) -> None:
        print(f"Took: {time.time() - self._start:.5f}")


timer = Timer()


@get("/", before_request=timer.start, after_response=timer.end)
def hello_world() -> dict[str, Any]:
    """Hi"""
    return {"hello": "world"}


app = Starlite(route_handlers=[hello_world])
```
Hitting the route returns `{"detail":"TypeError(\"Timer.start() missing 1 required positional argument: '_'\")"}`. Which is because we explicitly unbind any instance handed to those parameters.

The example could be solved by using a higher order function to wrap the method calls, but the aesthetics of that aren't that great.

You can assign callables to attributes of already existing instances without any binding problems, e.g.:

```python
class A: ...
a = A()
a.z = lambda: None
a.z()  # no worries
```

It's only when assigning a function to a class body that the binding becomes a problem:
```python
class Z:
   a = lambda: None
Z().a()  # TypeError: Z.<lambda>() takes 0 positional arguments but 1 was given
```

Even assigning a method to a class var on the controller would work properly without the unbinding:
```python
class A:
    def b(self):
        print(self)

class B:
    c = A().b

# note `self` not rebound to `__main__.B object`
B().c()  # <__main__.A object at 0x7f844e5c6b00>
```

The simplest change would be to raise an exception on startup if one of the lifecycle hook parameters is a method - at least it would fail fast. However, I think it is a shame we lose the ability to pass methods to these hooks in all cases when the only time it is actually a problem is when the handlers are normal functions assigned on the controller class body.

Maybe we could try and limit the unbinding to inside `Controller.__init__()`, I haven't tested this but something like:

```python
class Controller:
    ...
    def __init__(self, ...) -> None:
        ...
        if self.after_request and not ismethod(type(self).after_request):
            self.after_request = ensure_unbound(self.after_request)
            ...
```
Other (more nuclear) options...

We could deprecate `before_request`, `after_request` and `after_response` and add collection parameters that allow more than one handler per hook type, e.g., `before_request_handlers = [...]` where each handler is called in sequence and in that example, the first handler to return a response would short circuit the chain. This would prevent binding of funcs to the class instances as they'd be wrapped in the collection type.

Another option might be to introduce a `ControllerConfig` object.. so instead of assigning to classvars directly on `Controller` subclasses, it would look something like this:

```python
class AppController(Controller):
    config = ControllerConfig(path="/", before_request=...)
```

**Comment by Goldziher** [comment]
i agree that we should support methods, there is really no reason not to do this. You wanna do this?

**Comment by peterschutt** [comment]
Yep I'll do it in #280
